utils/bounded_workers.go

```diff
@@ -0,0 +1,81 @@
+// (c) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package utils
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+)
+
+type BoundedWorkers struct {
+	workerSpawn        chan struct{}
+	workerCount        atomic.Int32
+	outstandingWorkers sync.WaitGroup
+
+	outstandingWork sync.WaitGroup
+
+	work      chan func()
+	workClose sync.Once
+}
+
+// NewBoundedWorkers returns an instance of [BoundedWorkers] that
+// will spawn up to [max] goroutines.
+func NewBoundedWorkers(max int) *BoundedWorkers {
+	return &BoundedWorkers{
+		work:        make(chan func()),
+		workerSpawn: make(chan struct{}, max),
+	}
+}
+
+// startWorker creates a new goroutine to execute [f] immediately and then keeps the goroutine
+// alive to continue executing new work.
+func (b *BoundedWorkers) startWorker(f func()) {
+	b.workerCount.Add(1)
+	b.outstandingWorkers.Add(1)
+
+	go func() {
+		defer b.outstandingWorkers.Done()
+
+		if f != nil {
+			f()
+			b.outstandingWork.Done()
+		}
+		for f := range b.work {
+			f()
+			b.outstandingWork.Done()
+		}
+	}()
+}
+
+// Execute the given function on an existing goroutine waiting for more work, a new goroutine,
+// or return if the context is canceled.
+//
+// Execute must not be called after Wait.
+func (b *BoundedWorkers) Execute(ctx context.Context, f func()) bool {
+	b.outstandingWork.Add(1)
```

Comment:
If we don't track outstandingWork, we can close work while some caller may still be waiting on it.